pkg/pillar/zedcloud/deferred.go

```diff
@@ -24,97 +27,149 @@ import (
+	priorityCheckFunctions = append(priorityCheckFunctions, func(obj interface{}) bool {
+		return true
+	})
 	zedcloudCtx.deferredCtx = DeferredContext{
-		deferredItems: make(map[string]deferredItemList),
-		ticker:        flextimer.NewRangeTicker(longTime1, longTime2),
+		lock:                   &sync.Mutex{},
+		ticker:                 flextimer.NewRangeTicker(longTime1, longTime2),
+		sentHandler:            sentHandler,
+		priorityCheckFunctions: priorityCheckFunctions,
+		zedcloudCtx:            zedcloudCtx,
 	}
 	return zedcloudCtx.deferredCtx.ticker.C
 }
 
-// Try to send all deferred items. Give up if any one fails
+// HandleDeferred try to send all deferred items (or only one if sendOne set). Give up if any one fails
 // Stop timer if map becomes empty
 // Returns true when there are no more deferred items
-func HandleDeferred(zedcloudCtx *ZedCloudContext, event time.Time, spacing time.Duration) bool {
+func HandleDeferred(zedcloudCtx *ZedCloudContext, event time.Time, spacing time.Duration, sendOne bool) bool {
 
-	return zedcloudCtx.deferredCtx.handleDeferred(zedcloudCtx.log, event, spacing)
+	return zedcloudCtx.deferredCtx.handleDeferred(zedcloudCtx.log, event, spacing, sendOne)
 }
 
 func (ctx *DeferredContext) handleDeferred(log *base.LogObject, event time.Time,
-	spacing time.Duration) bool {
+	spacing time.Duration, sendOne bool) bool {
+	ctx.lock.Lock()
+	defer ctx.lock.Unlock()
+
+	if len(ctx.deferredItems) == 0 {
+		return true
+	}
 
-	log.Functionf("HandleDeferred(%v, %v) map %d\n",
+	log.Functionf("HandleDeferred(%v, %v) items %d",
 		event, spacing, len(ctx.deferredItems))
+
+	//sort items by time the entry was added/created in ascending order
+	sort.Slice(ctx.deferredItems, func(i, j int) bool {
+		t1 := ctx.deferredItems[i].lastModifiedTime
+		t2 := ctx.deferredItems[j].lastModifiedTime
+		return t1.Before(t2)
+	})
+
 	iteration := 0 // Do some load spreading
-	for key, l := range ctx.deferredItems {
-		log.Functionf("Trying to send for %s items %d\n", key, len(l.list))
-		failed := false
-		for i, item := range l.list {
-			if item.buf == nil {
-				continue
-			}
-			if item.buf.Len() == 0 {
-				log.Errorf("Zero length defered item for %s",
-					key)
-				continue
-			}
-			log.Functionf("Trying to send for %s item %d data size %d\n",
-				key, i, item.size)
-			resp, _, _, err := SendOnAllIntf(item.zedcloudCtx, item.url,
-				item.size, item.buf, iteration, item.bailOnHTTPErr)
-			if item.bailOnHTTPErr && resp != nil &&
-				resp.StatusCode >= 400 && resp.StatusCode < 600 {
-				log.Functionf("HandleDeferred: for %s ignore code %d\n",
-					key, resp.StatusCode)
-			} else if err != nil {
-				log.Functionf("HandleDeferred: for %s failed %s\n",
-					key, err)
-				failed = true
-				break
+	done := false
+	for _, f := range ctx.priorityCheckFunctions {
+		for ind := 0; ind < len(ctx.deferredItems); ind++ {
+			key := ctx.deferredItems[ind].key
+			item := ctx.deferredItems[ind]
+			if f(item.itemType) {
```

Comment:
Since there is no else branch, I find the code easier to read with an inverted test and continue.